File: cachatore/canales_pausados.py
"""Pausa temporal de canales — espejo Python del helper que vive en
functions/src/canales_pausados.ts y whatsapp-bot/src/canales_pausados.js.

Lee el doc META/canales_pausados con cache 5 min. Sirve para que el
cachatore (avisar_turno / enviar_resumen_diario_turnos) pregunte
"¿está pausado el canal X?" antes de encolar.

Diseño defensivo: cualquier falla de Firestore → "no hay pausa" (False).
"""
import time
from datetime import datetime, timezone

import choferes

COL_META = "META"
DOC_CANALES_PAUSADOS = "canales_pausados"
TTL_SEG = 5 * 60

_cache: dict | None = None
_cache_expira: float = 0.0


def _cargar() -> dict:
    global _cache, _cache_expira
    if _cache is not None and time.monotonic() < _cache_expira:
        return _cache
    try:
        db = choferes._db()
        snap = db.collection(COL_META).document(DOC_CANALES_PAUSADOS).get()
        _cache = snap.to_dict() or {} if snap.exists else {}
        _cache_expira = time.monotonic() + TTL_SEG
        return _cache
    except Exception:
        return _cache or {}


def esta_canal_pausado(key: str) -> bool:
    """`True` si `key` está pausado (con o sin fecha de fin, pero no
    vencida). `False` si no está en el doc, o si la fecha ya pasó.
    """
    mp = _cargar()
    raw = mp.get(key)
    if not isinstance(raw, dict):
        return False
    hasta = raw.get("hasta_iso")
    if isinstance(hasta, str) and hasta:
        try:
            # Acepta tanto ISO con Z como con offset.
            iso = hasta.replace("Z", "+00:00")
            dt_hasta = datetime.fromisoformat(iso)
            if dt_hasta.tzinfo is None:
                dt_hasta = dt_hasta.replace(tzinfo=timezone.utc)
            if datetime.now(timezone.utc) >= dt_hasta:
                return False
        except Exception:
            # Si la fecha no parsea, conservador: tratamos como pausa
            # indefinida (la app valida formato al escribir).
            pass
    return True


def invalidar_cache() -> None:
    global _cache, _cache_expira
    _cache = None
    _cache_expira = 0.0

File: whatsapp-bot/src/canales_pausados.js
// =============================================================================
// CANALES PAUSADOS — espejo Node del helper que vive en
// functions/src/canales_pausados.ts.
// =============================================================================
//
// Lee el doc META/canales_pausados con cache 5 min. Sirve para que los
// crons del bot (serviceDiario, vencimientosProximosConsolidado) y la
// alerta de cola creciente puedan preguntar "¿está pausada la categoría
// X?" antes de encolar.
//
// Diseño defensivo: cualquier falla → "no hay pausa" (devuelve false).
// Preferimos enviar de más a silenciar un mensaje crítico por un fallo
// de red.

const admin = require('firebase-admin');

let _db = null;
let _cache = null;
let _cacheExpiraMs = 0;
const TTL_MS = 5 * 60 * 1000;

function _getDb() {
  if (_db) return _db;
  _db = admin.firestore();
  return _db;
}

async function _cargar() {
  if (_cache && Date.now() < _cacheExpiraMs) return _cache;
  try {
    const snap = await _getDb()
      .collection('META')
      .doc('canales_pausados')
      .get();
    _cache = snap.exists ? (snap.data() || {}) : {};
    _cacheExpiraMs = Date.now() + TTL_MS;
    return _cache;
  } catch (_e) {
    return _cache || {};
  }
}

/**
 * Devuelve `true` si el canal `key` está pausado (no vencido).
 * Úsalo al inicio de un cron:
 *
 *     if (await estaCanalPausado('serviceDiario')) return;
 */
async function estaCanalPausado(key) {
  const map = await _cargar();
  const raw = map[key];
  if (!raw || typeof raw !== 'object') return false;
  const hasta = raw.hasta_iso;
  if (typeof hasta === 'string' && hasta.length > 0) {
    const ms = Date.parse(hasta);
    if (Number.isFinite(ms) && Date.now() >= ms) return false;
  }
  return true;
}

function invalidarCache() {
  _cache = null;
  _cacheExpiraMs = 0;
}

module.exports = {
  estaCanalPausado,
  invalidarCache,
};
